src/server.rs

```diff
@@ -72,11 +73,21 @@ where
     }
 }
 
+/// Given a host and port, creates a string of the form "host:port" or
+/// "[ho:st]:port", depending on whether the host is an IPv6 literal.
+fn join_host_port(host: &str, port: u16) -> String {
+    if let Ok(ip) = host.parse::<IpAddr>() {
+        format!("{}\0", SocketAddr::new(ip, port))
+    } else {
+        format!("{}:{}\0", host, port)
```

Comment:
Are you sure? Do we need to treat unparsable IP as invalid input?

Reply:
So I didn't read the comments above.
Nevermind.

Reply:
The host can be a domain name, eg, "google.com".